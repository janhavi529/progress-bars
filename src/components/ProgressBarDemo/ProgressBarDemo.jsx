import React, { Component } from 'react';
import axios from 'axios';
import Loader from 'react-loader-spinner';
import ProgressBar from './ProgressBar/ProgressBar';
import Button from '../form/Button/Button';
import Dropdown from '../form/Dropdown/Dropdown';
import '../../styles/scss/components/_progress-bar-demo.scss';

class ProgressBarDemo extends Component {
  state = {
    progressBars: [],
    buttons: [],
    limit: 100,
    error: false,
    loading: true,
    selectedBar: 1,
  }

  componentDidMount() {
    let updatedProgressBars = [];

    // Call the API to fetch progress bar data.
    axios.get('/bars').then(({ data }) => {
      if (data) {
        const { bars, buttons, limit } = data;

        // Store Progress Bar objects in state for business logic implementation.
        updatedProgressBars = bars.map((value, index) => ({
          barId: index + 1,
          value,
          label: 'Progress Bar #' + (index + 1),
        }));

        this.setState({
          progressBars: updatedProgressBars,
          buttons,
          limit,
          loading: false
        });
      }
    }).catch((error) => {
      this.setState({ error: error });
    });
  }

  render() {
    let progressBarList = [], buttonList = [], dropdownValues = [];
    const { loading, error, progressBars, buttons, limit, selectedBar } = this.state;

    if (loading) {
      return <Loader type="ThreeDots" color="grey" height={80} width={80} className="center-align"/>;
    } else if (error) {
      return <div className="error">Error occurred while loading API data: {error}</div>;
    } else {
      progressBarList = progressBars && progressBars.map(({ barId, value, label }) => {
        // Create dropdown menu options for progress bars.
        dropdownValues.push({
          id: barId,
          label,
        });

        return <ProgressBar key={barId} value={value} limit={limit} />;
      });

      buttonList = buttons.map((offset) => <Button key={offset} className="btn btn-light m-1" value={parseInt(offset)} onClick={(event) => this.onPercentageChange(event)} />);
    }

    return (
      <section className="bar-progress-wrapper">
        <header className="page-header">
          <h1 className="bar-progress-title">Progress Bar Demo</h1>
        </header>
        <main className="page-main">
          <div className="bar-progress-container">
            {progressBarList}
          </div>
          <div className="bar-progress-controls">
            <Dropdown label="Select Progress Bar" value={selectedBar} options={dropdownValues} onChange={(event) => this.onProgressBarSelect(event)} />
            <div className="mx-2 text-center"><div className="mb-1">Increment/Decrement Percentage</div>{buttonList}</div>
          </div>
        </main>
      </section>
    );
  }

  /* On every selection from the dropdown menu, set the selected progress bar. */
  onProgressBarSelect = (event) => {
    this.setState({ selectedBar: parseInt(event.target.value) });
  }

  /* Update selected progress bar on increasing/decreasing the bar percentage. */
  onPercentageChange = (event) => {
    const { progressBars, selectedBar } = this.state;
    const bars = progressBars && progressBars.reduce((accumulator, progressBar) => {
      if (progressBar.barId === selectedBar)
        progressBar.value = parseInt(progressBar.value) + parseInt(event.target.value);
      accumulator.push(progressBar);
      return accumulator;
    }, []);

    this.setState({
      progressBars: bars
    });
  }
}

export default ProgressBarDemo;
