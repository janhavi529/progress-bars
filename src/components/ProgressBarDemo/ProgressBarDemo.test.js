import React from 'react';
import { shallow, mount } from 'enzyme';
import ProgressBarDemo from './ProgressBarDemo';
import axios from 'axios';

describe('ProgressBarDemo', () => {
  it('Default ProgressBarDemo render', () => {
    const wrapper = shallow(<ProgressBarDemo />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('ProgressBarDemo - API Success', () => {
    it('Fetches API data', done => {
      const promise = new Promise((resolve) =>
        setTimeout(
          () =>
            resolve({
              data: {
                buttons: [49, 29, -37, -34],
                bars: [86, 27, 39],
                limit: 210
              },
            }),
          100
        )
      );

      axios.get = jest.fn(() => promise);
      const wrapper = mount(<ProgressBarDemo />);

      expect(wrapper.instance().state.progressBars.length).toEqual(0);
      promise.then(() => {
        wrapper.update();
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.instance().state.progressBars.length).toEqual(3);
        axios.get.mockClear();
        done();
      });
    });
  });

  describe('ProgressBarDemo - API Failure', () => {
    it('Doesn\'t render progress bar data when API data fetch fails', async (done) => {
      const promise = new Promise((reject) =>
        setTimeout(
          () => reject({ error: 'Unable to fetch data from API' }),
          100
        ));

      axios.get = jest.fn(() => promise);
      const wrapper = mount(<ProgressBarDemo />);

      expect(wrapper.instance().state.progressBars.length).toEqual(0);
      promise.then(() => {
        wrapper.update();
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.instance().state.progressBars.length).toEqual(0);
        axios.get.mockClear();
        done();
      });
    });
});

  it('Updates progress bar on percentage change', () => {
    const wrapper = shallow(<ProgressBarDemo />);
    wrapper.instance().setState({
      progressBars: [{
        barId: 1,
        value: 65,
        label: 'Progress Bar #1',
      },
      {
        barId: 2,
        value: 15,
        label: 'Progress Bar #2',
      },
      {
        barId: 3,
        value: 100,
        label: 'Progress Bar #3',
      }
      ],
      selectedBar: 2
    });

    wrapper.instance().onPercentageChange({
      target: {
        value: 25
      }
    });

    expect(wrapper.instance().state.progressBars[1].value).toEqual(40);
  });
  it('Updates selected progress bar on dropdown selection', () => {
    const wrapper = shallow(<ProgressBarDemo />);

    wrapper.instance().onProgressBarSelect({
      target: {
        value: 3
      }
    });

    expect(wrapper.instance().state.selectedBar).toEqual(3);
  });
});