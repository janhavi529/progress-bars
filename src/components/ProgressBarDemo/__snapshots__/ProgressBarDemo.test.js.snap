// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBarDemo Default ProgressBarDemo render 1`] = `ShallowWrapper {}`;

exports[`ProgressBarDemo ProgressBarDemo - API Failure Doesn't render progress bar data when API data fetch fails 1`] = `ReactWrapper {}`;

exports[`ProgressBarDemo ProgressBarDemo - API Success Fetches API data 1`] = `ReactWrapper {}`;
